File: Java - AWT - Swing/asteroids/Laser.java
package asteroids;

import java.util.*;

class Laser extends Entity {

    private boolean enemy;

    private Laser() {
    }

    static public void spawn(float x, float y, float heading, boolean enemy) {
        Queue<Laser> pool = new LinkedList<>();

        Laser asset = (pool.size() == 0 || pool.peek().active) ? new Laser() : pool.remove();
        asset.anim = (enemy) ? Astro.animations.get("laser_saucer") : Astro.animations.get("laser");
        asset.x = x;
        asset.y = y;
        asset.heading = heading;
        asset.dx = 6.0f * (float) Math.cos(heading);
        asset.dy = 6.0f * (float) Math.sin(heading);
        asset.active = true;
        asset.enemy = enemy;

        pool.add(asset);
        Astro.entities.add(0, asset);
    }

    static public void spawn(float x, float y, float heading) {
        spawn(x, y, heading, false);
    }

    public void update() {
        super.update();
        if (x < 0 || x > Astro.WIDTH || y < 0 || y > Astro.HEIGHT)
            active = false; // delete the laser pulse when it goes off the screen
    }

    public void check_collisions() {
        if (enemy)
            return; // collisions between enemy laser and ship handled by the ship

        for (Entity e : Astro.currentEntities) {
            active = active && !e.collides_with(this);
            if (!active)
                break;
        }
    }

    public boolean collides_with(Entity e) {
        if (enemy && !(e instanceof Laser) && intersects(e)) {
            active = false; // enemy laser gets destroyed by collision with ship
            return true;
        }
        return false;
    }
}

File: CSharp - Monogame SDK 3.7/Laser.cs
using System;
using System.Collections.Generic;

namespace Asteroids {

    public class Laser : Entity {

        bool enemy;

        private Laser() { }

        public static void Spawn(double x, double y, double heading, bool enemy = false) {
            Queue<Laser> pool = new Queue<Laser>();

            Laser laser = (pool.Count == 0 || pool.Peek().active) ? new Laser() : pool.Dequeue();
            laser.anim = (enemy) ? Astro.Instance.animations["laser_saucer"] : Astro.Instance.animations["laser"];
            laser.x = x;
            laser.y = y;
            laser.heading = heading;
            laser.enemy = enemy;
            laser.dx = 6 * Math.Cos(heading);
            laser.dy = 6 * Math.Sin(heading);
            laser.active = true;

            pool.Enqueue(laser);
            Astro.Instance.entities.Insert(0, laser);
        }

        override public void Update() {
            base.Update();
            if (x < 0 || x > Astro.Instance.WIDTH || y < 0 || y > Astro.Instance.HEIGHT)
                active = false; // delete the laser pulse when it goes off the screen
        }

        override public void CheckCollisions() {
            if (enemy) return; // collisions between enemy laser and ship handled by the ship

            foreach (Entity e in Astro.Instance.currentEntities) {
                active = active && !e.CollidesWith(this);
                if (!active) break;
            }
        }

        public override bool CollidesWith(Entity e) {
            if (enemy && e.GetType() != typeof(Laser) && Intersects(e)) {
                active = false; // enemy laser gets destroyed by collision with ship
                return true;
            }
            return false;
        }
    }
}
